onadata/apps/api/viewsets/connect_viewset.py

```diff
@@ -42,8 +44,15 @@ def user_profile_w_token_response(request, status):
     try:
         user_profile = request.user.profile
     except UserProfile.DoesNotExist:
-        user_profile, _ = UserProfile.objects.get_or_create(
-            user=request.user)
+        user_profile = cache.get(
+            f'{USER_PROFILE_PREFIX}{request.user.username}')
+        if not user_profile:
+            user_profile, __ = UserProfile.objects.get_or_create(
+                user=request.user)
+
+    cache.set(
+        f'{USER_PROFILE_PREFIX}{request.user.username}',
+        user_profile)
```

Comment:
I think we want to move this inside the `if` block after line 50 and 51 when we have retrieved the `user_profile` object. As it is at the moment, we are setting the cache value every time this function is called.

Reply:
Yeah, moved that. Thanks